File: day24/day24.cpp
#include <fstream>
#include <iostream>
#include <vector>

#include <fmt/format.h>

using namespace std;

using Component = pair<int,int>;

static vector<pair<int,int>> load(istream& input)
{
	vector<pair<int,int>> comps;
	int a, b;
	while (true)
	{
		input >> a;
		input.get();
		input >> b;
		if (!input)
		{
			break;
		}
		comps.emplace_back(make_pair(a, b));
	}
	return comps;
}

static size_t maxstrength;
static size_t maxlen;
static size_t maxlenstrength;

static void backtrack(size_t strength, size_t len, int free, vector<pair<int,int>>& comps)
{
	if (maxlen <= len)
	{
		maxlen = len;
		if (maxlenstrength < strength)
		{
			maxlenstrength = strength;
		}
	}
	if (maxstrength < strength)
	{
		maxstrength = strength;
	}

	for (size_t i = 0; i < comps.size(); i++)
	{
		auto t = comps[i];
		if (t.first != free && t.second != free)
		{
			continue;
		}
		int other = t.first == free ? t.second : t.first;
		comps.erase(comps.begin() + i);
		backtrack(strength + free + other, len+1, other, comps);
		comps.insert(comps.begin() + i, t);
	}
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}
	auto comps = load(input);
	input.close();

	backtrack(0, 0, 0, comps);
	cout << fmt::format("Part1: {}", maxstrength) << endl
	     << fmt::format("Part2: {}", maxlenstrength) << endl;
	return 0;
}

File: day24/day24.py
#!/usr/bin/env python3

import sys

class Component(object):
    def __init__(self, txt):
        self.ends = sorted(map(int, txt.split("/")))
        self.score = sum(self.ends)

    def __repr__(self):
        return repr(self.ends)

def load_components(txt):
    return [Component(c) for c in txt.splitlines()]

maxlen = 0
maxlenstrength = 0
maxstrength = 0
def backtrack(strength, length, free, comps):
    global maxstrength, maxlen, maxlenstrength
    if maxlen <= length:
        maxlen = length
        if maxlenstrength < strength:
            maxlenstrength = strength
    if maxstrength < strength:
        maxstrength = strength

    for i in range(len(comps)):
        if free in comps[i].ends:
            c = comps.pop(i)
            other = (free == c.ends[0]) and c.ends[1] or c.ends[0]
            backtrack(strength + free + other, length + 1, other, comps)
            comps.insert(i, c)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    c = load_components(txt)
    backtrack(0, 0, 0, c)
    print("Part1:", maxstrength)
    print("Part2:", maxlenstrength)
